cfme/middleware/deployment.py

```diff
@@ -43,91 +56,103 @@ class MiddlewareDeployment(MiddlewareBase):
+        deployments = MiddlewareDeployment.deployments(provider=haw_provider,server=ser_instance)
 
     """
     property_tuples = [('name', 'name')]
 
-    def __init__(self, name, provider=None, **kwargs):
+    def __init__(self, name, provider=None, server=None, **kwargs):
         if name is None:
             raise KeyError("'name' should not be 'None'")
+        if server is None or not isinstance(server, MiddlewareServer):
+            raise KeyError("'server' should be an instance of MiddlewareServer")
         self.name = name
+        self.server = server
         self.provider = provider
         self.id = kwargs['id'] if 'id' in kwargs else None
-        self.server = kwargs['server'] if 'server' in kwargs else None
 
     @classmethod
-    def deployments(cls, provider=None):
+    def deployments(cls, provider=None, server=None):
         deployments = []
-        if provider:
-            # if provider instance is provided try to navigate provider's deployments page
-            # if no deployments are registered in provider, returns empty list
-            if not provider.load_all_provider_deployments():
-                return []
-        else:
-            # if provider instance is not provided then navigates  to all deployments page
-            sel.force_navigate('middleware_deployments')
+        _get_deployments_page(provider=provider, server=server)
         if sel.is_displayed(list_tbl):
-            for page in paginator.pages():
+            _server = server
+            for _ in paginator.pages():
                 for row in list_tbl.rows():
-                    deployments.append(MiddlewareDeployment(name=row.deployment_name.text,
-                                                            server=row.server.text))
+                    if not server:
+                        _server = MiddlewareServer(provider=provider, name=row.server.text)
+                    deployments.append(MiddlewareDeployment(provider=provider, server=_server,
+                                                            name=row.deployment_name.text))
         return deployments
 
     @classmethod
     def deployments_in_db(cls, server=None, provider=None):
         deployments = []
-        rows = _db_select_query(server=server, provider=(provider.name if provider else None)).all()
+        rows = _db_select_query(server=server.name if server else None,
+                                provider=(provider.name if provider else None)).all()
+        _server = server
         for deployment in rows:
+            if not server:
+                _server = MiddlewareServer(name=deployment[3], provider=provider)
             deployments.append(MiddlewareDeployment(id=deployment[0], name=deployment[1],
-                                                    server=deployment[3], provider=deployment[4]))
+                                                    server=_server, provider=provider))
```

Comment:
here the idea is to construct object fully from db fields, to be validated later